update ds/bst (#36)
OI-wiki/i18n: docs/ds/bst.md

## Introduction to Binary Search Tree

Binary search tree is a tree-shape data structure of binary tree, which is defined as follows:

1. The empty tree is a binary search tree.

> 2. If the left subtree of the binary search tree is not empty, the accumulated weights of all points on the left subtree are less than the value of its root node.

Author: 改成了`weight sum of ..`? 查了下很少有用accumulate的 😅 🤔

> 2. If the left subtree of the binary search tree is not empty, the accumulated weights of all points on the left subtree are less than the value of its root node.

Reviewer: 感觉有什么误解 🌚 
这里好像不是加和
$w_{\text{t}} < w_{\text{root}}$ for all nodes t in the left subtree
我感觉是这个意思（

3. If the right subtree of the binary search tree is not empty, the accumulated weights of all points on the right subtree are greater than the value of its root node.

4. The left and right subtrees of the binary search tree are both binary search trees.

The time spent on basic operations on a binary search tree is proportional to the height of the tree. For a binary search tree with $n$ nodes, the optimal time complexity of these operations is $O(\log n)$ and the worst is $O(n)$ . The expected height of a binary search tree that is randomly built is $O(\log n)$ .

## Basic operations

> In the following code, we assume that $n$ is the number of nodes, $h$ is the height, `val[x]` is the value stored at node $x$, `cnt[x]` is the number of occurrences of the value stored in $x$ , and `lc[x]` and `rc[x]` are the left and right child nodes of node $x$ respectively.

Reviewer: ```suggestion
In the following code, we assume that $n$ is the number of nodes, $h$ is the height, `val[x]` is the value stored at node $x$, `cnt[x]` is the number of occurrences of the value stored in $x$ , and `lc[x]` and `rc[x]` are the left and right child nodes of node $x$, respectively.
```

}
```

### Find the ranking of elements

The ranking is defined as the number of numbers before the first identical element after sorting the array elements $+1$ .

> Maintain the subtree size $siz$ of each root node. To find the ranking of an element, first move from the root node to this element. If you move to the right, the answer is the number of left child node of the root plus the number of duplicates of the current node, and the size of the left child subtree at the end point $+ 1$ .

Reviewer: ```suggestion
Maintain the subtree size $\text{siz}$ of each root node. To find the ranking of an element, first move from the root node to this element. If you move to the right, the answer is the number of left child node of the root plus the number of duplicates of the current node, and the size of the left child subtree at the end point $+ 1$ .
```
第二句这样翻译有亿点乱，建议
```md
ans is the sum of the followings plus 1:
- a
- b 
- c
```

}
```

> ### Find the kth element

Reviewer: ```suggestion
### Find the k-th element
```

In a subtree, the ranking of the root node depends on the size of its left subtree.

If the size of the left subtree is greater than or equal to $k$ , then the element is in the left subtree;

> If the size of the left subtree is in the interval $[k-cnt,k-1]$ ( $cnt$ is the number of occurrences of the value of the current node), then this element is the root node of the subtree;

Reviewer: ```suggestion
If the size of the left subtree is in the interval $[k-\text{cnt},k-1]$ , where $cnt$ is the number of occurrences of the value of the current node, then this element is the root node of the subtree;
```

> If the size of the left subtree is less than $k-cnt$ , then the element is in the right subtree.

Reviewer: ```suggestion
If the size of the left subtree is less than $k-\text{cnt}$ , then the element is in the right subtree.
```

Time complexity is $O(h)$ .

```cpp
int querykth(int o, int k) {
  if (siz[lc[o]] >= k) return querykth(lc[o], k);
  if (siz[lc[o]] < k - cnt[o])
    return querykth(rc[o], k - siz[lc[o]] - cnt[o] + 1);
  return val[o];

>   // To find the node corresponding to the kth element, just return o directly

Reviewer: ```suggestion
  // To find the node corresponding to the k-th element, just return o directly
```
